### PR TITLE
Make history trie creation invalidate cache more aggresively (#RCHAIN…

### PR DESCRIPTION
**Pull request by dzajkowski**
…-3735)

## Overview
<sup>_What this PR does, and why it's needed_</sup>

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>
https://rchain.atlassian.net/browse/RCHAIN-3735

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Review by EncodePanda — APPROVED** [review]

**Comment by dzajkowski** [comment]
bors try

File: rspace/src/main/scala/coop/rchain/rspace/history/HistoryInstances.scala [hunk 1]
```diff
@@ -385,7 +385,7 @@ object HistoryInstances {
         _ <- historyStore.put(elements)
       } yield (
         elements.lastOption.map(Trie.hash).getOrElse(currentRoot),
-        elements.lastOption.map(t => (remainingPath, t))
+        elements.lastOption.map(t => (remainingPath, t)).orElse(previousModificationOpt)
```
Comment (ArturGajowy) [inline rspace/src/main/scala/coop/rchain/rspace/history/HistoryInstances.scala:388]: turns out widely passed accumulators are just as bad as global state? :D

**Review by ArturGajowy — APPROVED** [review]

File: rspace/src/test/scala/coop/rchain/rspace/history/HistorySpec.scala [hunk 2]
```diff
@@ -110,6 +110,11 @@ object TestData {
   val _zerosOnes: KeyPath       = (List.fill(16)(0) ++ List.fill(16)(1)).map(_.toByte)
   val _31zeros: KeyPath         = List.fill(31)(0).map(_.toByte)
   def zerosAnd(i: Int): KeyPath = _31zeros :+ i.toByte
+  def infixWithZeros(f: Byte, s: String): KeyPath = {
```
Comment (lukasz-golebiewski) [inline rspace/src/test/scala/coop/rchain/rspace/history/HistorySpec.scala]: `prefixWithZeros(s: String)` would be enough for your case

File: rspace/src/test/scala/coop/rchain/rspace/history/CachingHistorySpec.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    *  |
+    * S(28x0)
+    *  |
+    * PB(0, 1, 2)
+    *  | (2)
+    * S(000)
+    *  |
+    * Leaf (leafs aren't cached)
+  **/
+  "processing of multiple mixed actions" should "remove data form cache" in withEmptyTrie {
```
Comment (lukasz-golebiewski) [inline rspace/src/test/scala/coop/rchain/rspace/history/CachingHistorySpec.scala]: *from cache

File: rspace/src/test/scala/coop/rchain/rspace/history/CachingHistorySpec.scala [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    val caching      = CachingHistoryStore(historyStore)
+    val history      = new MergingHistory[Task](History.emptyRootHash, caching)
+    (history, caching)
+  }
+
+  protected def withEmptyTrie(
+      f: (MergingHistory[Task], CachingHistoryStore[Task]) => Task[Unit]
+  ): Unit = {
+    val (history, caching) = create
+    f(history, caching).flatMap(_ => history.close()).runSyncUnsafe(20.seconds)
```
Comment (lukasz-golebiewski) [inline rspace/src/test/scala/coop/rchain/rspace/history/CachingHistorySpec.scala]: Could you extract `runSyncUnsafe` to a common place for all the tests? Or at least for this one and others could be updated on demand

File: rspace/src/main/scala/coop/rchain/rspace/history/HistoryInstances.scala (hunk 1 above)
Comment (dzajkowski) [reply rspace/src/main/scala/coop/rchain/rspace/history/HistoryInstances.scala:388]: :sobbing_quietly:

**Review by lukasz-golebiewski — APPROVED** [review]

**Pull request merged**